### PR TITLE
Staged JIT expressions

### PR DESCRIPTION
**Pull request by PalumboN**
Fix on staged expression to compile `extB` bytecode :)

There are also:

- Some fixes for Constraints inference (with staged values)
- Avoid unnecessary Moves (due to staged expressions)

## TODO
- [x] Keep access to VMMemoryMap as JIT expressions
- [x] Compile a DruidVM that does not fail

## For next iteration
- [x] Fix `gen_ExtUnconditionalJump` -> `annotateBytecode:` https://github.com/Alamvic/druid/pull/86
- [ ] Fix write barrier in `storeAndPopReceiverVariableBytecode` https://github.com/Alamvic/druid/pull/87
- [x] Improve `generatePath` for JIT expressions (now are ignored, but registers can have constraints)
- [x] Fix staged loads
- [ ] Check why is it using more registers for some bytecodes (?

### CODE REVIEW

**Comment by PalumboN** [comment]
This is ready to be merged (if the CI approves)! Then we can continue working on the rest tasks from here.

File: Druid-Tests/DruidTestRTLCompiler.class.st [hunk 1]
```diff
@@ -138,6 +149,33 @@ DruidTestRTLCompiler >> gen_emptyPrimitiveWithArguments [
...
+			^ 0 ].
+		r54 := extB.
+		extB := r54 << 8 + byte1.
+		r115 := numExtB.
+		numExtB := r115 + 1.
+		^ 0 ].
+	r54 := extB.
+	extB := r54 << 8 + byte1.
+	r115 := numExtB.
+	numExtB := r115 + 1.
```
Comment (guillep) [inline Druid-Tests/DruidTestRTLCompiler.class.st:175]: betterish!

File: Druid-Tests/DruidTestRTLCompiler.class.st [hunk 2]
```diff
@@ -155,17 +193,13 @@ DruidTestRTLCompiler >> gen_extPushIntegerBytecode [
...
-	| t0 r43 currentBlock r5 live |
+	| live r5 r43 currentBlock |
 	live := 0.
 	r5 := extA.
 	extA := 0.
-	t0 := self allocateRegNotConflictingWith: live ifNone: [ ^ self unknownBytecode ].
-	live := live bitOr: (self registerMaskFor: t0).
-	self MoveCq: 1 R: t0.
-	true=BytecodeSetHasDirectedSuperSend ifTrue: [
+	BytecodeSetHasDirectedSuperSend=true ifTrue: [
```
Comment (guillep) [inline Druid-Tests/DruidTestRTLCompiler.class.st:200]: looks nice too!

File: Druid-Tests/DruidTestRTLCompiler.class.st [hunk 3]
```diff
@@ -200,44 +234,119 @@ DruidTestRTLCompiler >> gen_extSendSuperBytecode [
...
+	(byte1 + (r3 << 8))<0 ifTrue: [
+		| r11 jump1 t0 t1 |
+		r11 := numExtB.
+		extB := 0.
+		numExtB := 0.
+		((byte1 + (r3 << 8)) + (r11 * 2))>=0 ifTrue: [
+			self Jump:
+				(self ensureFixupAt: bytecodePC + 1 + (byte1 + (r3 << 8)) + 1).
+			deadCode := true.
+			^ 0 ].
```
Comment (guillep) [inline Druid-Tests/DruidTestRTLCompiler.class.st:253]: yes!

File: Druid-Tests/DruidTestRTLCompiler.class.st [hunk 4]
```diff
@@ -3636,153 +3820,152 @@ DruidTestRTLCompiler >> gen_shortUnconditionalJump [
...
-	jumpNext := self Jump: 0.
-	jumpTrue jmpTarget: self Label.
-	self genMoveConstant: 1 R: t2.
-	jumpNext jmpTarget: self Label.
-	currentBlock := self Label.
-	jump2 jmpTarget: currentBlock.
-	self ssFlushStack.
-	self CmpCq: 1 R: t2.
+	self AndCq: objectMemory getMemoryMap getSpaceMaskToUse R: t2.
+	self CmpCq: objectMemory getMemoryMap getOldSpaceMask R: t2.
```
Comment (guillep) [inline Druid-Tests/DruidTestRTLCompiler.class.st:3856]: nais!

File: Druid/DRAbstractMemoryAddress.class.st [hunk 5]
```diff
@@ -99,3 +99,9 @@ DRAbstractMemoryAddress >> rtlPushOperandQualifierForInstruction: aPush [
 DRAbstractMemoryAddress >> type [
 	^ type
 ]
+
+{ #category : #accessing }
+DRAbstractMemoryAddress >> type: aDRType [
+
+	type := aDRType
```
Comment (guillep) [inline Druid/DRAbstractMemoryAddress.class.st:106]: An address has a type?

File: Druid/DRBytecodeCompilerCompiler.class.st [hunk 6]
```diff
@@ -60,8 +60,9 @@ DRBytecodeCompilerCompiler >> initialize [
 		optimisations add:
 			(DRDeadCodeElimination then: DRRedundantCopyElimination).
 		optimisations add: DRCleanControlFlow new.
-		"optimisations add:
-			(DRDeadEdgeSplitting then: DRDeadBranchElimination)" ].
+		optimisations add: DRTailDuplicateStagedExpressions new.
```
Comment (guillep) [inline Druid/DRBytecodeCompilerCompiler.class.st:63]: gud

File: Druid/DRBytecodeIRGenerator.class.st [hunk 7]
```diff
@@ -308,21 +308,18 @@ DRBytecodeIRGenerator >> interpretPushWith: aRBMessageNode [
 DRBytecodeIRGenerator >> interpretRememberWith: aRBMessageNode [
 
 	| obj |
-	obj := (self visitOperand: aRBMessageNode arguments first)
-		       simpleConstantFold.
-
+	obj := self visitOperand: aRBMessageNode arguments first.
+	"	
+	cogit 	backEnd saveAndRestoreLinkRegAround: [cogit CallRT: ceStoreCheckTrampoline].
```
Comment (guillep) [inline Druid/DRBytecodeIRGenerator.class.st:313]: comment?

File: Druid/DRJitCompileTimeExpression.class.st [hunk 8]
```diff
@@ -15,6 +15,36 @@ DRJitCompileTimeExpression class >> expression: anExpression [
 		yourself
 ]
 
+{ #category : #accessing }
+DRJitCompileTimeExpression class >> getNewSpaceMask [
+
+	^ self expression: 'objectMemory getMemoryMap getNewSpaceMask'
```
Comment (guillep) [inline Druid/DRJitCompileTimeExpression.class.st:21]: Would be nice to generalise this later

**Review by guillep — APPROVED** [review]
Some small comments

File: Druid/DRBytecodeIRGenerator.class.st (hunk 7 above)
Comment (PalumboN) [reply Druid/DRBytecodeIRGenerator.class.st:313]: The comment is the code to generate (so why `saveLinkReg`)

File: Druid/DRAbstractMemoryAddress.class.st (hunk 5 above)
Comment (PalumboN) [reply Druid/DRAbstractMemoryAddress.class.st:106]: That seams :P

I don't remember, I had broken tests.

Anyway, as an address could be, for example, an interpreter reference, that could be an object pointer (like `trueObj`) or an address (like `oldSpaceStart`).

**Pull request merged**